src/main/java/org/cryptacular/CiphertextHeaderFactory.java

```diff
@@ -0,0 +1,344 @@
+   *
+   * @return  Decoded header.
+   *
+   * @throws  EncodingException  when ciphertext header cannot be decoded.
+   * @throws  StreamException  on stream IO errors.
+   */
+  private static V1CiphertextHeader decodeV1(final InputStream input) throws EncodingException, StreamException
+  {
+    final int length = ByteUtil.readInt(input);
+    if (length < 0) {
+      throw new EncodingException("Bad ciphertext header");
+    }
+
+    final byte[] nonce;
+    int nonceLen = 0;
+    try {
+      nonceLen = ByteUtil.readInt(input);
+      if (nonceLen > V1CiphertextHeader.MAX_NONCE_LEN) {
+        throw new EncodingException("Bad ciphertext header: maximum nonce size exceeded");
+      }
+      nonce = new byte[nonceLen];
+      input.read(nonce);
+    } catch (ArrayIndexOutOfBoundsException e) {
+      throw new EncodingException("Bad ciphertext header");
+    } catch (IOException e) {
+      throw new StreamException(e);
+    }
+
+    String keyName = null;
+    if (length > nonce.length + 8) {
+      final byte[] b;
+      int keyLen = 0;
+      try {
+        keyLen = ByteUtil.readInt(input);
+        if (keyLen > V1CiphertextHeader.MAX_KEYNAME_LEN) {
+          throw new EncodingException("Bad ciphertext header: maximum key length exceeded");
+        }
+        b = new byte[keyLen];
+        input.read(b);
+      } catch (ArrayIndexOutOfBoundsException e) {
+        throw new EncodingException("Bad ciphertext header");
+      } catch (IOException e) {
+        throw new StreamException(e);
+      }
+      keyName = new String(b);
+    }
+
+    return new V1CiphertextHeader(nonce, keyName);
+  }
+
+
+  /**
+   * Read <code>output.length</code> bytes from the input stream into the output buffer.
+   *
+   * @param  input  Input stream.
+   * @param  output  Output buffer.
+   *
+   * @throws  StreamException  on stream IO errors.
+   */
+  private static void readInto(final InputStream input, final byte[] output)
+  {
+    try {
+      input.read(output);
+    } catch (IOException e) {
+      throw new StreamException(e);
+    }
+  }
+
+
+  /**
+   * Read a single byte from the input stream.
+   *
+   * @param  input  Input stream.
+   *
+   * @return  Byte read from input stream.
+   */
+  private static byte readByte(final InputStream input)
+  {
+    try {
+      return (byte) input.read();
+    } catch (IOException e) {
+      throw new StreamException(e);
+    }
+  }
+
+
+  /**
+   * Determines if two byte array ranges are equal bytewise.
+   *
+   * @param  a  First array to compare.
+   * @param  aOff  Offset into first array.
+   * @param  b  Second array to compare.
+   * @param  bOff  Offset into second array.
+   * @param  length  Number of bytes to compare.
+   *
+   * @return  True if every byte in the given range is equal, false otherwise.
+   */
+  private static boolean arraysEqual(final byte[] a, final int aOff, final byte[] b, final int bOff, final int length)
```

Comment:
This doesn't seem like a method we need to implement. Nothing in `java.util.Arrays` will work here?